### ISSUE TITLE
HTTPIP and HTTPPort unexpectedly interpolated as "<no value>"

### ISSUE THREAD

**Issue by jsjohns** [mitchellh/packer#1890]
I'm attempting to create a CentOS virtual machine using the `vmware-iso` builder. I have supplied a value for the `http_directory` option and see that a HTTP server is started successfully on the host.

The `HTTPIP` and `HTTPPort` variables are not being set correctly. The boot line that is passed to the VM ends up being `linux console=tty0 ks=http://<no value>:<no value>/anaconda-ks.cfg<wait5><enter>`.

Tested on:

- Packer v0.7.5, v0.7.2
- Windows 8.1, Ubuntu 14.10
- VMware Workstation 11, VMware player 6.0.5

Template:

```json
{
    "builders": [
      {
        "type": "vmware-iso",
        "iso_url": "file:///X:/linux.iso",
        "iso_checksum": "556DF073D184CB260B9D10A94D136F65C7541F66",
        "iso_checksum_type": "sha1",
        "ssh_username": "packer",
        "ssh_wait_timeout": "20m",
        "shutdown_command": "shutdown -P now",
        "http_directory": "X:\\Temporary\\Packer\\http",
        "http_port_min" : 9001,
       "http_port_max" : 9001,
        "boot_command": "linux console=tty0 ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/anaconda-ks.cfg<wait5><enter>",
        "disk_size": "50000",
        "guest_os_type": "linux",
        "vmx_data": {
          "numvcpus": "2",
          "memsize": "2048"
        }
      }
    ]
}
```

**Comment by rickard-von-essen** [comment]
Please rerun with [debug logging](http://www.packer.io/docs/other/debugging.html) and gist the log.

**Comment by jsjohns** [comment]
I tried to enable debug logging, but I'm getting an error with that as well. Roughly:

```
set PACKER_LOG="1"
set PACKER_LOG_PATH="C:\packer-test"
packer.exe build test.json
Couldn't setup log output: open "C:\packer-test": The filename, directory name, or volume label syntax is incorrect.
```

**Comment by jsjohns** [comment]
Here's what I was able to capture by copy/pasting from the command prompt window: https://gist.github.com/jsjohns/f5abf2350a8313fe63bc

**Comment by rickard-von-essen** [comment]
@rasa Does debug logging work on Windows? How should it be enabled? Maybe we have to clarify the documentation if the above is not correct.

This is really strange:
> Found file 'C:/ProgramData/VMware/vmnetnat.conf'
> Host IP for the VMware machine: 192.168.246.1
> Typing the boot command over VNC...

By reading the code it looks like everytihing is fine when the IP for the host have been found (line 2). But then the value doesn't seem to be set in the template (```boot_command```).

Do you get the exact same error on Ubuntu?

**Comment by tompsweeney** [comment]
I'm having the same issue using `virtualbox-iso`. 

Host: Windows 8.1
Packer Version: 0.7.5
VirtualBox Version: 4.3.22

Here is my template:
```
{
	 "builders": [
		 {
		  "type": "virtualbox-iso",
		  "guest_os_type": "RedHat_64",
		  "iso_url": "http://mirror.csclub.uwaterloo.ca/centos/7.0.1406/isos/x86_64/CentOS-7.0-1406-x86_64-Minimal.iso",
		  "iso_checksum": "e3afe3f1121d69c40cc23f0bafa05e5d",
		  "iso_checksum_type": "md5",
		  "ssh_username": "packer",
		  "ssh_password": "packer",
		  "http_directory": "http",
		  "output_directory": "build",
		  "boot_command": "<esc>linux ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/anaconda-ks.cfg<enter>",
		  "ssh_wait_timeout": "30s",
		  "shutdown_command": "echo 'packer' | sudo -S shutdown -P now"
		}
	]
}
```

Here is a link to the gist of my log: [packer.log](https://gist.github.com/tompsweeney/e143b2bdfe3f2eaeb251)

**Comment by michael-os** [comment]
Hi, I encounter the very same problem, running Packer 0.7.5 on Fedora 21, trying to generate a virtualbox-iso image. Has anyone been able to solve or workaround this issue?

**Comment by tompsweeney** [comment]
I haven't found a solution, but my workaround was to hard-code the IP and port. On my system, virtualbox always seems to use 10.0.2.2:9001 for the HTTP server.

**Comment by michael-os** [comment]
Thanks for reporting back, @tompsweeney!

@rickard-von-essen could you remove the waiting-reply tag, since tompsweeney posted a rather extensive log almost 4 months ago. Would be great to get the attendance of some maintainer on this topic.

**Comment by stephen-mw** [comment]
It looks like it comes down to a serialization problem. Make sure your packerfile doesn't contain newlines in the boot command string. It should be one giant string.

**Comment by rickard-von-essen** [comment]
I have reproduced this on Windows 8.1. I'll add some info when gathered some more info.

**Comment by ajmazurie** [comment]
This is not a Windows-specific bug; I have the exact same issue on OSX 10.10.2 with the following template: http://pastebin.com/83imtAdm

Note that both builders (for VMWare and VirtualBox) have this issue; i.e., none of the two virtualization environment is provided with value for the `HTTPIP` nor `HTTPPort` variables. I tried various values for the `http_directory parameter` (e.g., local directory, sub-directory, parent directory) to no avail.

This is my first foray into Packer, and so far a roadblock. Let me know if there is a way to bypass this issue.

**Comment by dragon788** [comment]
@ajmazurie So basically if you try and one-line the boot command string it doesn't handle the HTTPIP correctly, but if you do it on multiple lines in an array it works? This would explain why I haven't seen this myself, as I copied a template that had the command on multiple lines broken up with \n between each piece.

**Comment by rickard-von-essen** [comment]
This is fixed in Packer v 0.8.0 just released. If anyone still see this please open a new issue.

**Comment by jsjohns** [comment]
I don't see this issue referenced in the changelog. Should it be there?

**Comment by bish0polis** [comment]
FWIW I was able to get this to reproduce on 086.

Then I made it go away.  What made it break?

foo.json included 
"boot_command": [
                "<tab> text noverifyssl ks={{user `ks_baseurl`}}/{{user `ks_urldir`}}/{{user `ks_name`}}<enter><wait>"
:
    "variables": {
        "ks_baseurl": "http://{{ .HTTPIP }}:{{ .HTTPPort }}",
:

And moving it to a more classic setup got it working again:
:
"boot_command": [
                "<tab> text noverifyssl ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/{{user `ks_urldir`}}/{{user `ks_name`}}<enter><wait>"
:

... but I'll miss the added versatility.  I haven't seen this week's release, but if it's not already a feature in there, I'd love to see it!  

(yes, I know noverifyssl is pointless when the URL is expanded to just an http://)

(Sorry if I jacked the thread: this is where I came to find out about MY issue, and I'm hoping at least someone else may be as dumb as me but come this way searching)

 - bish..
